### PR TITLE
Feature creation node

### PR DESCRIPTION
**Pull request by Dianeod**

### CODE REVIEW

File: code/nodes/dataPreprocessing/funcs.q [hunk 1]
```diff
@@ -91,3 +91,7 @@ dataPreprocessing.nullEncode:{[feat;func]
    flip 0^(func each flip feat)^flip[feat],names!nullValues
    ]
   }
+
+
+// Temporary infreplace function until toolkit is update
+dataPreprocessing.infreplace:{.ml.infreplace["f"$x]}
```
Comment (cmccarthy1) [inline code/nodes/dataPreprocessing/funcs.q]: I think this function should probably use the infinity replace outlined in the current master version of AutoML i.e. at https://github.com/cmccarthy1/automl-1/blob/master/code/updml.q

File: code/nodes/featureCreation/nlp/featureCreate.q [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  charPrep:featureCreation.nlp.proc[feat;cfg;0b;(::)];
+  // Table returned with NLP feature creation, any constant columns are dropped
+  featNLP:charPrep`feat;
+  // run normal feature creation on numeric datasets and add to nlp features if relevant
+  cols2use:cols[feat]except charPrep`stringCols;
+  if[0<count cols2use;
+    nonTextFeat:charPrep[`stringCols]_feat;
+    featNLP:featNLP,'featureCreation.normal.create[nonTextFeat;cfg]`features
+    ];
+  featureExtractEnd:.z.T-featExtractStart;
```
Comment (cmccarthy1) [inline code/nodes/featureCreation/nlp/featureCreate.q:25]: This should come after the drop constant to be consistent with the other methods

File: code/nodes/preprocParams/preprocParams.q [hunk 3]
```diff
@@ -2,11 +2,11 @@
...
 \d .automl
 
-inputKeys :`config`dataDescription`creationTime`sigFeats`symEncode`symMap
-inputTypes:"!+tSS!"
+inputKeys :`config`dataDescription`creationTime`sigFeats`symEncode`symMap`featModel
+inputTypes:"!+tSS!<"
 preprocParams.node.inputs  :inputKeys!inputTypes
 preprocParams.node.outputs :"!"
-preprocParams.node.function:{[cfg;descrip;ctime;sigfeat;symenc;symmap]
+preprocParams.node.function:{[cfg;descrip;ctime;sigfeat;symenc;symmap;featmodel]
```
Comment (cmccarthy1) [inline code/nodes/preprocParams/preprocParams.q:9]: Function needs to have the 'pre-function description' added styled similar to the featureDescription.q node

File: code/nodes/featureCreation/normal/funcs.q [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+// @kind function
+// @category featureCreation
+// @fileoverview Used in the recursive application of functions to a kdb+ table 
+// @param feat {tab} The feature data as a table 
+// @param func {(lambda;str)} function to be applied to the table
+// return {table} with the desired transforms applied recursively
+featureCreation.normal.applyFunc:{[feat;func]
+  typ:type func;
+  func:$[-11h=typ;get func;100h=typ;func;.automl.featureCreation.normal.default];
```
Comment (cmccarthy1) [inline code/nodes/featureCreation/normal/funcs.q:15]: May be worth adding a comment over this to say that once Deanna's changes to tts/sigfeats are in then this function should use the same 'util.qpyFuncSearch' function being used in that PR s.t. python feature creation procedures could in theory also be produced

File: code/nodes/featureCreation/normal/funcs.q [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  func feat
```
Comment (cmccarthy1) [inline code/nodes/featureCreation/normal/funcs.q]: Should check that the return from this is a table, if it's not attempt to run .ml.df2tab, if it's still not a simple table error out

**Review by cmccarthy1 — CHANGES_REQUESTED** [review]
Some minor additions to commenting, code ordering and a note on future addition depending on https://github.com/cmccarthy1/automl-1/pull/15 to be added

File: code/nodes/preprocParams/preprocParams.q (hunk 3 above)
Comment (Dianeod) [reply code/nodes/preprocParams/preprocParams.q:9]: This node isn't completed yet, and "pre-function description" is usually added by owner of node (Deanna in this case)

Comment (cmccarthy1) [reply code/nodes/preprocParams/preprocParams.q:9]: Noted, didn't take into account it was a different node than the featureCreation

**Review by cmccarthy1 — APPROVED** [review]

**Pull request merged**